python/tskit/trees.py

```diff
@@ -5073,6 +5076,71 @@ def alignments(self, *, reference_sequence=None, missing_data_character=None):
             a[site_pos] = h
             yield a.tobytes().decode("ascii")
 
+    @property
+    def individual_populations(self):
+        """
+        Returns the length-``num_individuals`` array containing, for each
+        individual, the ``population`` attribute of their nodes, or
+        ``tskit.NULL`` for individuals with no nodes. Errors if any individual
+        has nodes with inconsistent non-NULL populations.
+        """
+        if self._individual_populations is None:
+            pops = np.repeat(tskit.NULL, self.num_individuals)
+            for n in self.nodes():
```

Comment:
What about doing it this way?
```python
for ind in self.individuals():
     if len(ind.nodes) > 0: 
          ind_pops = set(self.node(u).population for u in ind.nodes)
          if len(ind_pops) > 1:
               raise ValueError("...")
           pops[ind.id] = ind_pops.pop() # or whatever the set operation is, can't remember
```            
This avoids iteration over all nodes, and makes the "populations must be exactly equal" semantics easier to implement. 

I wouldn't worry too much about the efficiency of doing set operations here, as we'll probably want to drop this functionality down to the C api at some point anyway. I'm pretty sure I've written the equivalent C code in msprime, so we could just copy that.

Reply:
That'd work, but I was writing this to be translatable to C more easily (which I was planning to do in this PR; if these operations are slow they'll be a bottleneck for things I do!)

Reply:
This is totally translatable to efficient C - I've done it all already [here](https://github.com/tskit-dev/msprime/blob/c11a2f12c72dd054a0ade0767474184ceec8281c/lib/msprime.c#L7207) in msprime.